### ISSUE TITLE
A more precise method of determining if a file is definitely a virus

### ISSUE THREAD

**Issue by PPPDUD** [Piggo4676/Foss-AV#12]
You see, so far, Foss AV can approximate if a virus is in a file by checking if the lines or filenames are suspicious. However, this mechanism is extremely unreliable and prone to false positives. 

Additionally, viruses can use formats like exe files, where it's not a good idea to check the "lines" of the  
file. 
A better alternative for files that are known to not be text files, such as .exe and .app files, is to use SHA256 hashes (see https://stackoverflow.com/questions/22058048/hashing-a-file-in-python).

You can obtain these hashes and check them against the hashes of known-bad malware, allowing for you to determine if an executable is a virus.

**Comment by jdev082** [comment]
I recommended something like this earlier on. We should definitely have a sha256 database with the hashes of malware samples. This is similar to how other AVs do things. We can have this database in a different repository and have it included in the main repository as a submodule.

**Comment by Piggo4676** [comment]
Mmm, yes, that was an issue I was thinking about, and I agree that sha256 hashes would be a better way to do this. the way I was doing it was just an example or a prototype. also the "functional" folder is just for all scripts that fully work on there own btw.

**Comment by Piggo4676** [comment]
what about worms.

**Comment by ninjamar** [comment]
> Additionally, viruses can use formats like exe files, where it's not a good idea to check the "lines" of the file. A better alternative for files that are known to not be text files, such as .exe and .app files, is to use SHA256 hashes (see https://stackoverflow.com/questions/22058048/hashing-a-file-in-python).

Inside `utils` is current a python package that can be imported and the `compsig` function can be used to compare hashes of files. I propose that we have some standard for contributions. Currently there aren't much code comments and the naming conventions are a little odd.

**Comment by Piggo4676** [comment]
We can use both of the methods (hashing/sha-256 AND file comparing using code flags)

**Comment by jdev082** [comment]
> We can use both of the methods (hashing/sha-256 AND file comparing using code flags)
> But for this, we need a MASSIVE and I mean **_MASSIVE_** database of virus sha-256 codes.

We should make this a Linux AV so there's less malware to have in the db.

**Comment by PPPDUD** [comment]
Worms are fundamentally the same as other malware. Just look for the hashes
of their files and delete them.

**Comment by PPPDUD** [comment]
> We can use both of the methods (hashing/sha-256 AND file comparing using code flags) But for this, we need a MASSIVE and I mean **_MASSIVE_** database of virus sha-256 codes.

What about a database of known _good_ software? Because I have truckloads of those.

**Comment by PPPDUD** [comment]
> > We can use both of the methods (hashing/sha-256 AND file comparing using code flags)
> > But for this, we need a MASSIVE and I mean **_MASSIVE_** database of virus sha-256 codes.
> 
> We should make this a Linux AV so there's less malware to have in the db.

No support. That's like trying to clean walls that are already clean. Target Windows instead, since it has a bigger malware problem, especially in China. [<small>source</source>](https://www.microsoft.com/en-us/wdsi/threats)

**Comment by jdev082** [comment]
> No support. That's like trying to clean walls that are already clean. Target Windows instead, since it has a bigger malware problem, especially in China. [<small>source</source>](https://www.microsoft.com/en-us/wdsi/threats)
Linux does have malware. While uncommon now, this may change in the near future.

**Comment by PPPDUD** [comment]
> > No support. That's like trying to clean walls that are already clean. Target Windows instead, since it has a bigger malware problem, especially in China. [source](https://www.microsoft.com/en-us/wdsi/threats)
> 
> Linux does have malware. While uncommon now, this may change in the near future.

Perhaps we should attempt to support as many NT and Unix-like systems as possible? After all, a new, non-Rosetta based ransomware has emerged on macOS, and we need to protect them.

**Comment by ninjamar** [comment]
> What about a database of known _good_ software? Because I have truckloads of those.

No - For obvious reasons

> No support. That's like trying to clean walls that are already clean. Target Windows instead, since it has a bigger malware problem, especially in China. [source](https://www.microsoft.com/en-us/wdsi/threats)

I don't use windows



> Perhaps we should attempt to support as many NT and Unix-like systems as possible? After all, a new, non-Rosetta based ransomware has emerged on macOS, and we need to stop it.

OSX and Linux? We could also add software to purge old files or show directories that occupy lots of storage.

**Comment by Piggo4676** [comment]
How about we target as many operating systems as we can? the more, the better. try and make it linux, windows and mac since those are the most common operating systems.

**Comment by jdev082** [comment]
> How about we target as many operating systems as we can? the more, the better. try and make it linux, windows and mac since those are the most common operating systems.

That's not something we can pull off.

**Comment by Piggo4676** [comment]
why not?

**Comment by Piggo4676** [comment]
Yeah. they trigger antiviruses but they don't actually run. It's a bit weird but yeah. If you dpwnload them, they're safe and they dont do anything, but antiviruses still go grrr when they see them.

**Comment by PPPDUD** [comment]
> How about we target as many operating systems as we can? the more, the better. try and make it linux, windows and mac since those are the most common operating systems.

Macs are not an operating system. They are computers and do not have to run macOS.

**Comment by Piggo4676** [comment]
> Macs are not an operating system. They are computers and do not have to run macOS.

Sorry, that's what I meant when I said Mac. I meant MacOS. my bad.

**Comment by PPPDUD** [comment]
It's okay :).
